[Lens] Refactor drag and drop

mbondyra opened the pull request:
## Summary

When I created drag and drop for Lens, the API I went for was not the most readable one. It was designed this way because I wanted to gain some performance, but it was very hard to maintain the performance gain with a lot of changes in the drag and drop area because all the pieces of the code needed to memoized in a tricky way and it wasn't communicated well.
In the end it works even without these tricks so I decided to simplify it in this PR.

The main changes include:

1. Instead of multiple `useState` per parameter, we keep all the state in reducer both for `ReorderProvider` and `RootDragDropProvider`. Thanks to that we get multiple improvements:
2. The code in `DragDrop` component becomes more descriptive as we don't run multiple state updates when user executes an action but one state update describing what actually happens (eg. `dispatchDnd({type: 'selectDropTarget' ....})`. The internal logic of the update lives in the reducer.
3. We don't have to pass `trackUiCounterEvents` as another prop to `DragDrop` and run it wherever we need - instead we pass it as a middleware to the context and run before dispatching (and it's very easy to add more middlewares if we need extra integrations at some point!)
4. We also run a11y announcements as a middleware instead of inside `DragDrop` component
5. The `ChildDragDropProvider` props look much cleaner:
    before:
```
      <ChildDragDropProvider
        keyboardMode={keyboardModeState}
        setKeyboardMode={setKeyboardModeState}
        dragging={draggingState.dragging}
        setA11yMessage={setA11yMessage}
        setDragging={setDragging}
        activeDropTarget={activeDropTargetState}
        setActiveDropTarget={setActiveDropTarget}
        registerDropTarget={registerDropTarget}
        dropTargetsByOrder={dropTargetsByOrderState}
        dataTestSubjPrefix={dataTestSubj}
        onTrackUICounterEvent={onTrackUICounterEvent}
      >
        {children}
      </ChildDragDropProvider>
```
after:
```
<ChildDragDropProvider value={[state, dispatch]}>{children}</ChildDragDropProvider>
```
6. Created custom hook `useDragDropContext` instead of using `useContext(DragContext)` and making DragContext private. This way we will avoid potential problems with using context outside of root.
7. Bonus thing - if we ever decide to move to redux, the structure is there already



What I am still not happy with is that the tests are very domain-dependant instead of user-driven - instead of checking the store actions, I should check the interface from the user perspective. I will try to work on it once I find some time between more important tasks though.

elasticmachine commented:
Pinging @elastic/kibana-visualizations @elastic/kibana-visualizations-external (Team:Visualizations)

mbondyra commented on packages/kbn-dom-drag-drop/src/providers/announcements.tsx:
```diff
@@ -10,11 +10,7 @@ import { i18n } from '@kbn/i18n';
-type AnnouncementFunction = (
-  draggedElement: HumanData,
-  dropElement: HumanData,
-  announceModifierKeys?: boolean
```
this param is actually not needed - it was added when the implementation was different. Basically it checks if we want to add the information to the user about the extra keys (eg 'shift to swap') and if the key is pressed (so swap is highlighted) then didn't inform the user about the keys. I think it's better to always inform the user about modifier keys so I removed it.

stratoula reviewed (commented):
This looks good to me, I am not an dnd expert but I like the simplification of the api. @jughosta can you also take a look here please?

I wonder why a simple change has increased the async bundle of event annotations 🤔

mbondyra commented:
@stratoula I managed to fix the bundle issue. The problem was caused by moving a11y announcements to the RootDragDropProvider, which made the bundle size bigger. After digging deeper, I realized that we were unnecessarily importing the provider into the annotations plugin and surrounding annotations with it. I removed this import from annotation plugin. To simplify things I also decided to change directly accessing the context `useContext(DragContext)` but using a custom hook that not only fetches the context but also checks if we're currently within its scope. This tweak should do the trick and prevent us from bugs difficult to catch. 😊

stratoula commented:
Wow @mbondyra, thank you so much for looking into it. These small nits make our products better!

stratoula reviewed (approved):
changes LGTM in case of green CI if @jughosta is also fine!

jughosta commented on src/plugins/event_annotation/public/components/group_editor_controls/annotation_list.tsx:
```diff
@@ -110,7 +108,7 @@ export const AnnotationList = ({
           getAdditionalClassesOnDroppable={
             DropTargetSwapDuplicateCombine.getAdditionalClassesOnDroppable
           }
-          dropTypes={dragging ? ['field_add'] : []}
```
Previously it was necessary for reorder styles to be applied only after drag started. Was this behaviour changed?

mbondyra replied:
Yes, it's from the PR before this one - the styles are applied inside the DragDrop component where we check for `dragging` so no need to do the check here as we only have one type of dropTypes for annotations.

drewdaemon replied:
I like it.

jughosta reviewed (approved):
Great work!

drewdaemon commented on src/plugins/event_annotation/public/get_table_list.tsx:
```diff
@@ -34,28 +33,26 @@ export const getTableList = (
   services: EventAnnotationListingPageServices
 ) => {
   return (
-    <RootDragDropProvider>
```
As discussed offline, this removal breaks the gated draggable dimension button interface (sorry—easy to miss 😅 ). We should fix this before merging!

drewdaemon commented:
It looks like something changed with the ghost:

Before this PR

<img width="448" alt="Screenshot 2023-07-10 at 2 44 04 PM" src="https://github.com/elastic/kibana/assets/315764/e35e9cf2-941b-4dd9-a7c3-bbfef03b168b">

After this PR

<img width="474" alt="Screenshot 2023-07-10 at 2 40 58 PM" src="https://github.com/elastic/kibana/assets/315764/197bc830-56b9-4bb7-9045-f987ac041056">

Could it be related to [this change](https://github.com/elastic/kibana/pull/161257/files#diff-df81e373d1392595c5944f0935c5aa2eb657dd5cf23e0f3a11b769b98df4b42bL113)?

drewdaemon reviewed (approved):
Would be nice to get that ghost problem fixed, but approving so as not to hold you up when I'm OOO!

mbondyra commented:
Thank you @drewdaemon, yep, it is the problem with this one indeed. I did fix it in another PR but actually seeing some other issues that it caused I reverted it. While checking I added reordering to annotation group editing here in this PR  - it's hidden under the feature var, so if there's any problem it's not visible for the users.
